File: booter_fc/flashcart_specifics/r4isdhc/r4isdhc.c
// This file is licensed CC0-1.0 by Taiju Yamada.

#include <stdio.h>
#include <string.h>

// arbitrary number. Don't know if this can be changed, because other converters of this type uses this value too... -lifehackerhansol
#define S 433264

#define BUFLEN (1<<19)

int main(int argc, char **argv) {
    int size = S - 0x450, u;
    if(argc < 3) {
        fprintf(stderr, "r4isdhc in out\n");
        return 1;
    }
    FILE* in = fopen(argv[1], "rb");
    if(!in) {
        fprintf(stderr, "cannot open\n");
        return 2;
    }
    FILE* out = fopen(argv[2], "wb");
    if(!out) {
        fclose(in);
        fprintf(stderr, "cannot open\n");
        return 3;
    }
    unsigned char buf[BUFLEN];
    memset(buf, 0, 0x450);
    buf[0] = 0x12, buf[1] = 0x01, buf[2] = 0x00, buf[3] = 0xea;
    buf[0xec] = 0x49, buf[0xed] = 0x00, buf[0xee] = 0x00, buf[0xef] = 0xeb;
    fwrite(buf, 1, 0x450, out);
    while(u = fread(buf, 1, BUFLEN, in)) fwrite(buf, 1, u, out), size-=u;
    fclose(in);
    memset(buf, 0, BUFLEN);
    while(size > 0) fwrite(buf, 1, size < BUFLEN ? size : BUFLEN, out), size -= BUFLEN;
    fclose(out);
    return 0;
}

File: booter_fc/flashcart_specifics/r4isdhc/r4isdhc.py
#
# ARM9 relocator
# Patch an ARM9 binary to change RAM location to 0x02000000
# and entrypoint to 0x02000450
# Inspired by Xenon++'s r4isdhc patcher
#
# Copyright (C) 2022-2023 DS-Homebrew
#
# SPDX-License-Identifier: 0BSD
#

from array import array
from os import stat
from sys import argv, exit

TARGET_SIZE = 0x69C70
HEADER_SIZE = 0x450


def main():

    argc = len(argv)
    if (argc < 3):
        print("r4isdhc.py <in> <out>")
        exit(1)

    buf = array("B")

    buf.extend([0] * HEADER_SIZE)
    buf[0x00] = 0x12
    buf[0x01] = 0x01
    buf[0x02] = 0x00
    buf[0x03] = 0xEA
    buf[0xEC] = 0x49
    buf[0xED] = 0x00
    buf[0xEE] = 0x00
    buf[0xEF] = 0xEB

    try:
        input_size = stat(argv[1]).st_size
        with open(argv[1], "rb") as i:
            buf.fromfile(i, input_size)
    except Exception:
        print("cannot open input")
        exit(2)

    remaining = TARGET_SIZE - HEADER_SIZE - input_size
    if remaining > 0:
        buf.extend([0] * remaining)

    try:
        with open(argv[2], "wb") as o:
            buf.tofile(o)
    except Exception:
        print("cannot open output")
        exit(3)

    print("Success.\n")
    exit(0)


main()
